### ISSUE TITLE
Develop latitude/longitude storage for tags

### ISSUE THREAD

**Issue by milaaraujo** [publiclab/plots2#4183]
Develop latitude/longitude storage for tags.
 
>"The lon/lat storage i think would be useful for a few reasons. We do proximity right now based on truncating the lat/lon strings, which is a bit crude and can have some failures -- i think this is documented in an issue comment, actually. But let's discuss there! Mainly, we can't order by a string the way we could by a decimal column." @jywarren

* [x] Add fields to node and populate fields, retroactively, with the tag - lat/lon - values (#4292)
* [x] Store lat/lon into node - we should be looking in the node.rb add_tag method (#4379)
* [ ] Synchronize deletion between the two means of storing lat/lon - tag controller
* [ ] Prototype ways to subscribe to new node posts based on proximity
* [ ] Update/refactor `tagNearbyNodes()` API method that uses lat/lon node tags to use lat/lon node fields

**Comment by milaaraujo** [comment]
Hey @jywarren, do you know which issue is this problem documented? I could not find it.

And I didn't understand what do you mean with the second request "prototype ways to subscribe to new node posts based on proximity". Can you give me some clarification?

**Comment by jywarren** [comment]
For your first Q, I think the issue is that our "nearby" logic in the API is based on truncating the tag so we get a lower resolution, and then finding all tags that start with that string -- a bit odd, here:

https://github.com/publiclab/plots2/blob/9dd50d1e1008bd23b13c1f21fd56e7c87d1f0fd8/app/services/search_service.rb#L106

Advantages are that we use the existing tag system, it's relatively simple, and since the length of the string is tied to location precision and privacy, we don't need to store more than the necessary precision (see this for more: https://github.com/publiclab/leaflet-blurred-location/)

I can't find the discussion but one big flaw of this is that two points like `10.9999` and `11.0001` are really very close, but the "prefix" method won't work to find that they are nearby.

The possible solution is to add new columns for longitude, latitude, and precision. We could add them to tags, and keep storing them separately, or we could add them to nodes. In either case we'll have to be sure that the string representation of a location (in a tagname) is synced with the decimal representation, and also the precision matches. 



### Proximity subscriptions

Basically this would adapt the existing subscription code but expand it so that you could subscribe to a location -- like your profile location (not sure how to select a location otherwise... could start with this but later generalize to anything with `.lat` and `.lon` methods?). When content gets posted, tagged within X distance of the subscribed location, you get a notification.

In theory we could poll "nearby content" via the API periodically and issue notifications. That could be easier. But we need some way to record that you are following a place. Hope that makes sense!

**Comment by jywarren** [comment]
For the "flaw" i describe above, we could start with a test, like copying this one: https://github.com/publiclab/plots2/blob/5b55750e2a8001ed98e629ad31a1e78f51a4c642/test/functional/api/search_api_test.rb#L121

basically we'd query `get '/api/srch/nearbyPeople?query=30.99,39.99` instead of `get '/api/srch/nearbyPeople?query=31.00,40.00'` and it would fail even though we are looking for a not much different location.

**Comment by milaaraujo** [comment]
Ok, let's start with the latitude/longitude storage!

>The possible solution is to add new columns for longitude, latitude, and precision. We could add them to tags, and keep storing them separately, or we could add them to nodes. In either case we'll have to be sure that the string representation of a location (in a tagname) is synced with the decimal representation, and also the precision matches.

In my opnion it makes more sense these new columns to tags, since it already stores the string representation of a location. What do you think? 

One more question... Where in the code do we get the location and save it as a string?

**Comment by milaaraujo** [comment]
I’m thinking about it and actually it makes more sense to store the lat/lon in the nodes!

**Comment by milaaraujo** [comment]
Hey @jywarren, after we're done with #4292 we will have the new fields (lat/lon/precision) added to the node table. In addition, the existing values in the tags (lon/lat) will be migrated to these new fields.

Then, I think the next step would be to change the code to save lat/lon to the new fields, right? Right now, wherein the code do we get the node location and save as tags?

**Comment by jywarren** [comment]
Hi, Camila. Currently lat long are entered through the location tag
interface which is initiated with the map marker icon button in the tagging
interface. So, perhaps upon doing node.add_tag(), we could intercept and
store true node location. But we would have to do this retroactively as
well, for existing tags - this could be done in a migration too.

Also, once location tags are deleted we should delete the corresponding
node tags. This sounds like a set of PRs we could break up in a checklist,
what do you think?

**Comment by milaaraujo** [comment]
> So, perhaps upon doing node.add_tag(), we could intercept and store true node location

-> Is it happening here (https://github.com/publiclab/plots2/blob/master/app/controllers/map_controller.rb#L158-L159), right? Here are we adding just the the node tags? Or the user tags as well? 

> But we would have to do this retroactively as well, for existing tags - this could be done in a migration too.

-> This is already done here: #4292! :)

> Also, once location tags are deleted we should delete the corresponding
node tags.

-> We have some methods using the node tags, first we need to update all these methods... Then we can delete the node tags.

**Comment by milaaraujo** [comment]
I will check `add_tag` method in the `node.rb` file then. Once we finish #4292 I will work on the next two itens of the checklist! :)

> Re deleting node tags, we want to keep using node tags, but we're just talking about deleting node tags at the moment that a given node's lat/lon is deleted, and vice versa - i.e. synchronizing deletion between the two means of storing lat/lon.

Ohhh, now I understood what you meant. Makes sense! I updated the checklist.

**Comment by jywarren** [comment]
This is looking great, thanks. Whenever we have to synchronize 2 forms of storage like this, it's a little bit worrisome, but I think that if we look at this thoroughly we should be OK. 

Note that @okonek has implemented the nearby nodes logic here using tags - should we still merge this and then refactor once we have decimal lat/lon storage? https://github.com/publiclab/plots2/pull/4128/

Thanks!

**Comment by jywarren** [comment]
Note the exact line where it does a `NodeTag` search here; that'd be the only are that'd have to change: https://github.com/publiclab/plots2/pull/4128/files#diff-9546dd6074cdca7a128c2bf04fa5367aR98

**Comment by jywarren** [comment]
Also, it's useful to link to the issue/PR numbers for the checklist, if they exist, to help tie everything together. Would you mind adding issue numbers where they exist, or even "related to #0000"? Thanks!

**Comment by milaaraujo** [comment]
After I finish the next two items in the list -  store lat/lon into node and synchronize deletion between the two means of storing lat/lon - we can open new issues to refactor all methods that use lat/lon tags. Actually, this is the last item of the checklist! So other collaborators can help with this and I can work on the other points of our contract, what do you think?

**Comment by jywarren** [comment]
That's super!
